elfloader-tool/src/arch-arm/smp_boot.c

```diff
@@ -34,7 +34,11 @@ void non_boot_main(void)
 #endif
     /* Spin until the first CPU has finished initialisation. */
     while (!non_boot_lock) {
-#ifndef CONFIG_ARCH_AARCH64
+#ifdef CONFIG_ARCH_AARCH64
+        /* The compiler may optimize this loop away, add a dsb()
```

Comment:
The comment seems incorrect, the variable is declared `volatile`, so the compiler is not allowed to optimize this out. However, the barrier might sill be needed to force a OO-pipeline not to try smart things. Instead of the barrier, we might also consider using https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html to let the compiler pick the appropriate meachnism.

Reply:
Yes you're correct it's declared `volatile` so this is not needed. I can add a `asm volatile("": : :"memory")` to the loop body to let the compiler handle this?

Reply:
cpu_idle() is just `dsb()` and `wfi()` paired, so we can make this common.

Reply:
I don't understand. `cpu_idle()` is doing something completely different, so this should not be used here.

Reply:
The current code as is
```
    while (!non_boot_lock) {
#ifdef CONFIG_ARCH_AARCH64
        dsb();
#else
        cpu_idle();
#endif
    }
```

And since `cpu_idle()` is itself: dsb() + wfi(), the dsb() would already be common.

Reply:
Ah ok, I see. But It's a bit odd that ARMv7 is doing a `wfi()` here also - I can't see who triggers the event on the other core that will bring us out of the `wfi` then. Did this ever work or was this stuck forever then - or did something else trigger an even by chance? I can't find anything in the commit history that explains this.

Reply:
Maybe this can be simplified to just a barrier and we can find a way to do a cache invalidation (not just aarch64) on `non_boot_lock` when the boot CPU finishes initializing?

Reply:
Since on aarch64 the D-Cache is disabled before the loop anyway, just a `dsb()` should be sufficient here. And it seems this worked well for your colleagues when the made the code change. 
Since I'm not really sure how this is supposed to work on ARMv7/AARCH32 and we can't test this now either, it might be best to just leave it and add a comment that calling `cpu_idle()` this looks like a hack that should either be explained or reworked.